src/orange/detail/BIHBuilder.cc

```diff
@@ -0,0 +1,228 @@
+Overload(Ts...) -> Overload<Ts...>;
+}  // namespace
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from vector of bounding boxes and storage for LocalVolumeIds.
+ */
+BIHBuilder::BIHBuilder(VecBBox bboxes,
+                       BIHBuilder::BboxStorage* bbox_storage,
+                       BIHBuilder::LVIStorage* lvi_storage,
+                       BIHBuilder::InnerNodeStorage* inner_node_storage,
+                       BIHBuilder::LeafNodeStorage* leaf_node_storage)
+    : bboxes_(bboxes)
+    , bbox_storage_(bbox_storage)
+    , lvi_storage_(lvi_storage)
+    , inner_node_storage_(inner_node_storage)
+    , leaf_node_storage_(leaf_node_storage)
+{
+    CELER_EXPECT(!bboxes_.empty());
+
+    centers_.resize(bboxes_.size());
+    std::transform(bboxes_.begin(),
+                   bboxes_.end(),
+                   centers_.begin(),
+                   [&](BoundingBox const& bbox) { return center(bbox); });
+
+    partitioner_ = BIHPartitioner(&bboxes_, &centers_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create BIH Nodes.
+ */
+BIHParams BIHBuilder::operator()() const
+{
+    VecIndices indices;
+    VecIndices inf_volids;
+
+    for (auto i : range(bboxes_.size()))
+    {
+        LocalVolumeId id(i);
+
+        if (!is_infinite(bboxes_[i]))
+        {
+            indices.push_back(id);
+        }
+        else
+        {
+            inf_volids.push_back(id);
+        }
+    }
+
+    VecNodes nodes;
+    this->construct_tree(indices, nodes, static_cast<BIHNodeId>(-1));
+
+    auto [inner_nodes, leaf_nodes] = this->arrange_nodes(std::move(nodes));
+
+    BIHParams params;
+
+    params.bboxes = ItemMap<LocalVolumeId, BoundingBoxId>(
+        make_builder(bbox_storage_).insert_back(bboxes_.begin(), bboxes_.end()));
+
+    params.inner_nodes
+        = make_builder(inner_node_storage_)
+              .insert_back(inner_nodes.begin(), inner_nodes.end());
+    params.leaf_nodes = make_builder(leaf_node_storage_)
+                            .insert_back(leaf_nodes.begin(), leaf_nodes.end());
+
+    params.inf_volids = make_builder(lvi_storage_)
+                            .insert_back(inf_volids.begin(), inf_volids.end());
+
+    return params;
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Recursively construct BIH nodes for a vector of bbox indices.
+ */
+void BIHBuilder::construct_tree(VecIndices const& indices,
+                                VecNodes& nodes,
+                                BIHNodeId parent) const
+{
+    auto current_index = nodes.size();
+    nodes.resize(nodes.size() + 1);
+
+    if (partitioner_.is_partitionable(indices))
+    {
+        BIHInnerNode node;
+        node.parent = parent;
+
+        auto p = partitioner_(indices);
```

Comment:
You could just make a local `BIHPartitioner partition` to this function, since it's not used anywhere else, and only called once for a set of bboces.